storm-server/src/main/java/org/apache/storm/scheduler/utils/IConfigLoader.java

```diff
@@ -0,0 +1,138 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.storm.scheduler.utils;
+
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.InputStreamReader;
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.util.Map;
+
+import org.apache.storm.Config;
+import org.apache.storm.DaemonConfig;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.yaml.snakeyaml.Yaml;
+import org.yaml.snakeyaml.constructor.SafeConstructor;
+
+public interface IConfigLoader {
+
+    /**
+     * Load scheduler configs
+     * It first retrieves parameters like the location of the scheduler config file from the conf;
+     * and then get the actual scheduler configs associated with the configKey from that location.
+     * @param configKey The key from which we want to get the scheduler config.
+     * @return The scheduler configs
+     */
+    Map<?,?> load(String configKey);
+
+    String SCHEDULER_CONFIG_LOADER_URI = "scheduler.config.loader.uri";
+    String SCHEDULER_CONFIG_LOADER_POLLTIME_SECS = "scheduler.config.loader.polltime.secs";
+    String SCHEDULER_CONFIG_LOADER_TIMEOUT_SECS = "scheduler.config.loader.timeout.secs";
+
+    int DEFAULT_POLLTIME_SECS = 600;
+    int DEFAULT_TIMEOUT_SECONDS = 10;
+
+    static IConfigLoader getConfigLoader(Map<String, Object> conf) {
+        Map loaderParams = (Map) conf.get(DaemonConfig.SCHEDULER_USER_POOLS_LOADER_PARAMS);
+        if (loaderParams != null) {
+            String uriString = (String) loaderParams.get(SCHEDULER_CONFIG_LOADER_URI);
+            if (uriString != null) {
+                try {
+                    URI uri = new URI(uriString);
+                    String scheme = uri.getScheme();
+                    switch (SchemeType.toSchemeType(scheme)) {
+                        case FILE:
+                            return new FileConfigLoader(loaderParams);
```

Comment:
To me it feels like this would be a great place to use a ServiceLoader instead of having a hard coded FileConfigLoader or ArtifactoryHttpConfigLoader.

Reply:
I have been thinking about using a ServiceLoader and tried so. But I realized that we actually want to have only one type of loader working in the mean time (one loader for one type of scheme). So I think `switch-case` here seems cleaner.   

The way I can think of applying ServiceLoader here is to have every type of loaders check the `scheme` in their `load()` function and then either directly return `null` (because they don't deal with this type of `scheme`) or return the result. Something like:

For HttpConfigLoader
```
Map load(loaderParams) {
    scheme = getScheme();
    if (!scheme.equalsIgnoreCase("http")) {
        return null;
    }
    //else: process and return result;
}
```
For ArtifactoryHttpConfigLoader,
```
Map load(loaderParams) {
    scheme = getScheme();
    if (!scheme.equalsIgnoreCase("artifactory+http")) {
        return null;
    }
    //else: process and return result;
}
```
Is this what you are talking about? Sorry if I mis-understood it.

Reply:
Actually I was thinking we might want to use a factory pattern so we can cache connections and other things in the IConfigLoader implementation.  Something perhaps like.

```java
public interface IConfigLoaderFactory {
    IConfigLoader createIfSupported(URI uri, Map<String, Object> conf);

    private static ServiceLoader<IConfigLoaderFactory> serviceLoader = ServiceLoader.load(IConfigLoaderFactory.class);
    
    public static IConfigLoader open(URI uri, Map<String, Object> conf) {
        for (IConfigLoaderFactory factory: serviceLoader) {
            IConfigLoaderFactory ret = factory.createIfSupported(uri, conf);
            if (ret != null) {
                return ret;
            }
        }
        throw some exception();
    }
}

public class HttpArtifactoryServiceLoaderFactory implements IConfigLoaderFactory{
    public IConfigLoader createIfSupported(URI uri, Map<String, Object> conf);
        if (!"artifactory+http".equals(uri.getScheme())) {
            return null;
        }
        return new ...;
    }
}
```

Reply:
This seems a great idea to me. I will try to do so. Thanks